# Lab 2.03 - Univariate Analysis


In [1]:
# Importing the necessary packages
import numpy as np                                  # "Scientific computing"
import scipy.stats as stats                         # Statistical tests

import pandas as pd                                 # Data Frame
from pandas.api.types import CategoricalDtype

import matplotlib.pyplot as plt                     # Basic visualisation
from statsmodels.graphics.mosaicplot import mosaic  # Mosaic diagram
import seaborn as sns                               # Advanced data visualisation

## Exercise 3 - Kitten litters

A vet keeps a record of the number of kittens in litters produced by cats in his practice. The code block below contains the data as a frequency table.

- First put the data into one list of individual observations: [1, 1, 2, 2, ... 5, 5, 6, 6, 6, 6, 7, 7, 8]  
- Visualise the data set using all suitable chart types that you know of.
- Calculate:
    - the median size of a litter
    - the mean size of a litter
    - the lower quartile
    - the upper quartile

In [2]:
kit_dict = {
    'Number in litter': [1, 2, 3,  4, 5, 6, 7, 8],
    'Frequency':        [2, 4, 7, 11, 8, 4, 2, 1]
    }

kittens = pd.DataFrame.from_dict(kit_dict)


In [3]:
# ft = [[x[0]] * x[1] for x in zip(kit_dict['Number in litter'], kit_dict['Frequency'])]
# r = []
# for s in ft:
#     for x in s:
#         r.append(x)

expanded = kittens.loc[kittens.index.repeat(kittens['Frequency']), 'Number in litter']

print(expanded)

0    1
0    1
1    2
1    2
1    2
1    2
2    3
2    3
2    3
2    3
2    3
2    3
2    3
3    4
3    4
3    4
3    4
3    4
3    4
3    4
3    4
3    4
3    4
3    4
4    5
4    5
4    5
4    5
4    5
4    5
4    5
4    5
5    6
5    6
5    6
5    6
6    7
6    7
7    8
Name: Number in litter, dtype: int64


In [ ]:
kittens = kittens.set_index('Number in litter')

In [ ]:
kittens.plot()

In [5]:
kitmean = expanded.mean(numeric_only=True)
kitmedian = expanded.median(numeric_only=True)
kitquantile = expanded.quantile([0.25,0.75])
print("Mean:\n", kitmean, "\n")
print("Median:\n", kitmedian, "\n")
print("Quantiles (25% and 75%):\n", kitquantile)

Mean:
 4.128205128205129 

Median:
 4.0 

Quantiles (25% and 75%):
 0.25    3.0
0.75    5.0
Name: Number in litter, dtype: float64


### Some answers:

- Median: 4.0
- Mean: 4.128205128205129
- Lower Quartile: 3.0
- Upper Quartile: 5.0

Source: Complete Probability & Statistics 1 for Cambridge International As & A Level